### PR TITLE
Comments: Add HTML toolbar for the edit textarea

### LINKED ISSUE #21663: Comments: add more formatting options in edit mode
We currently provide folks a plain text area, let's also add some formatting helpers like we do in wp-admin:

screen shot 2018-01-18 at 2 13 03 pm

screen shot 2018-01-18 at 2 18 39 pm

### PR DESCRIPTION
**Pull request by Copons**
Fix #21663
Depends on #22042

Add an HTML toolbar to help adding quick tags when editing comments in the Comments Management section.

screen shot 2018-02-01 at 12 07 16

## Testing instructions

Open `/comments` and click on Edit on some comments.
Use the new HTML toolbar to open and close tags.
Make sure the inserted tags feel "right" (e.g. the expected newlines or indents).

### Note

While testing this, it could happen that clicking on Edit doesn't keep the comment on screen.
It's a known bug, taken care in https://github.com/Automattic/wp-calypso/pull/22125.
Please disregard it here.

### CODE REVIEW

**Comment by gwwar** [comment]
@Copons the changeset here looks rather reasonable, I would probably take this over #22042. We don't add the risk of regressing the editor that way, and we always have the option to go back and perform the janitorial if we find that the feature sets between each editor is pretty stable.

I would call in more folks for testing this though, maybe lets ask flow patrol for help?

**Comment by Copons** [comment]
@gwwar Sure, if you think it's manageable to review, I'm all for this.
I'll do some last checks and fixes, and mark this to "needs review" asap.

**Comment by rachelmcr** [comment]
I tested out the live branch (in Chrome on Mac OS 10.13.3), comparing it to the behavior in WP Admin Comments and the behavior in the Calypso HTML editor. The new toolbar worked well both for adding formatting to existing text in the comment (highlighting the text and selecting a toolbar button) and adding new formatted text (opening a tag, entering text, closing the tag).

I saw a couple minor things that seemed off, but they aren't new/unique:

1. If I don't close an HTML tag, it isn't auto-closed for me. That is consistent with WP Admin Comments but it would be nice to help close tags if possible rather than having to rely on the `close tags` button. (In the Calypso editor, if you switch from HTML to Visual any open tags are closed for you.)
2. The `close tags` button doesn't work if you save the comment without closing all the HTML tags and then try to edit and close the tags later. This is also consistent with WP Admin Comments behavior but would be a nice to have.

**Comment by Copons** [comment]
> If I don't close an HTML tag, it isn't auto-closed for me.

@rachelmcr this is a very interesting suggestion, and it's quite easy to implement by simply triggering the Close Tags button on save: if there are open tags, we'll close them; otherwise we're just good to go.

Though, we should keep in mind that the Close Tags button only keeps track of the tags opened via the toolbar, so if there are tags manually opened by writing their HTML in the textarea, those would stay open regardless.
At the same time, if a tag was opened via the toolbar, and closed by manually writing the HTML, that tag is considered open by the Close Tags button, and therefore on save we'd have a duplicate closed tag.

Honestly, I'd say that the value of automatically attempting to close tags (which is a quick and dirty content sanitization) wins over the risks of duplicating closed tags (which results in an incorrect but safe HTML).
Though I'd rather have more opinions on this.

**Comment by bsessions85** [comment]
I am trying this on Mac Firefox 57.0.4 and the html tags aren't being added. I click them and they just bounce me to the end of the comment. I don't see any console errors or anything, it just doesn't work. Chrome and Safari both worked fine.

**Comment by hoverduck** [comment]
I also tried Firefox (Linux) and IE11 with no success :confused: 

On another note, is there a reason that it's only available on Editing comments, but not Replying?

**Comment by gwwar** [comment]
> On another note, is there a reason that it's only available on Editing comments, but not Replying?

I'd be open to adding that in a follow up, if we get a design 👍 .

**Comment by gwwar** [comment]
@Copons  re: closing tags. It's simpler to run a transform after save to close off malformed tags, versus having the tool keep that much state. We can always iterate in the future if this is still a friction point.

**Comment by brbrr** [comment]
Win 10 & MSEdge - was unable to add HTML tags.

While testing this feature in Chrome - I noticed 1 minor issue which seems not quite related to this particular PR: #22155

Q: is there any plans to add Preview? With HTML is not always easy to predict the output, so it would be nice to be able to check the preview before saving.

**Comment by Copons** [comment]
Dear testers, I think I've fixed whatever was wrong with FF, IE11, and Edge. Do please check again whenever you please. 🙇 

> If I don't close an HTML tag, it isn't auto-closed for me.

>  It's simpler to run a transform after save to close off malformed tags, versus having the tool keep that much state.

@rachelmcr @gwwar Let's keep this for a follow-up PR. This one is already huge by itself. :) 

> On another note, is there a reason that it's only available on Editing comments, but not Replying?

I've thought about it, but decided to not include such a thing in this PR, to avoid bloating it even more than how it currently is.

Should we reuse this toolbar, it shouldn't be that hard at all. It just needs a little bit of care in adapting the various moving parts of the Reply textarea.

> Q: is there any plans to add Preview? With HTML is not always easy to predict the output, so it would be nice to be able to check the preview before saving.

For this question I'd defer to @drw158.
Though, it must be said that Comments is not our team's current effort, so while we're happy to keep fixing bugs and polishing, maybe we're not up for such a big change right now.

**Comment by kwight** [comment]
Took a spin through with Mac'n'Firefox, everything worked well.

I'm still weirded out by the contrast of Reply and Edit: Reply drops down and visual context is easily kept. Edit, however, snaps up, covering the comment info and changing the size of the visual block, often leaving me suddenly unclear if I'm even looking at the right comment any more. But that's not here nor there for this PR 🙃

**Comment by bsessions85** [comment]
I tried again on Firefox and MS Edge and it is working now. The focus still goes to the end of the comment though, which is annoying.

**Comment by designsimply** [comment]
Tested with Firefox 58.0.1 on macOS 10.13.3 and I'll echo @bsessions85's comment that the cursor jumping to the end of the comment when trying to add HTML in the middle is pretty disruptive. I also had a bit of trouble with extra line returns getting added and some of the HTML (`ol` and `li`) wouldn't save:

<strong>Video:</strong> [2m6s](https://cloudup.com/c9EaLwaUe9m)
<sup>Seen at https://calypso.live/comments/all/design5279.wordpress.com using Firefox 58.0.1 on macOS 10.13.3.</sup>

**Comment by designsimply** [comment]
Also see https://github.com/Automattic/wp-calypso/issues/21246.

**Comment by gwwar** [comment]
> Let's keep this for a follow-up PR. This one is already huge by itself

Of course 😄  I suppose that wasn't clear from original comments.

**Comment by Copons** [comment]
> some of the HTML (ol and li) wouldn't save

@designsimply Couldn't reproduce this. :(

| Chrome | Firefox |
| --- | --- |
| feb-07-2018 11-16-22 | feb-07-2018 11-17-46 |

It's especially odd because we don't really do anything to the comment content, when we send it to the server.
It's the server then that does its sanitization magic, which should be the very same magic wp-admin uses—and I've simply added to the toolbar the same tags available on wp-admin.

Wonder if there is some other reason for that. 🤔

**Comment by Copons** [comment]
@bsessions85 @designsimply I've pushed a fix for the cursor position.
It was a logic error, so the solution works automatically for all browsers.
I've tested it on Chrome and Firefox on MacOS 10.13, and on Edge and IE11 on Win10.

---

I've also added a workaround that fixes an odd behaviour on Chrome.
When closing a dialog (it opens by clicking on the link and image buttons), the focus didn't return to the textarea as expected, but on the button.
It felt very flow-breaking to me, and very odd that it only happened on Chrome.
I couldn't figure out the actual reason, so the solution is not the most elegant: I've added a `textarea.focus()` on dialog close delayed by 500ms.
Considering the dialog's fade out animation takes 200ms, we've only left with a window of 300ms where we might end up stealing the focus back from the user.
I think that 300ms is long enough to be sure to focus back to the textarea, and quick enough to not risking stepping on the users' toes.

File: client/lib/html-toolbar/index.js [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+ * @param {String} before The content before the selection.
+ * @param {String} content The new content to insert.
+ * @param {String} after The content after the selection.
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertContent = ( textarea, before, content, after, onInsert = noop ) => {
+	const { selectionEnd, value } = textarea;
+	const newContent = before + content + after;
+
+	if ( Env.gecko ) {
```
Comment (gwwar) [inline client/lib/html-toolbar/index.js]: Let's avoid pulling in Env, and use `document.queryCommandSupported` instead. As a general note feature detection over browser detection is preferred.

https://developer.mozilla.org/en-US/docs/Web/API/Document/queryCommandSupported

File: client/lib/html-toolbar/index.js [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+		// always moves the cursor to the end of the content.
+		// A workaround involving manually setting the cursor position and inserting the editor content
+		// is needed to put the cursor back to the correct position.
+		textarea.value = newContent;
+		textarea.focus();
+		document.execCommand( 'insertText', false, content );
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		textarea.focus();
+		onInsert( newContent );
+	} else if ( !! Env.ie && Env.ie >= 11 ) {
```
Comment (gwwar) [inline client/lib/html-toolbar/index.js]: `if ( ! document.queryCommandSupported( 'insertText' ) ) {`

File: client/lib/html-toolbar/index.js [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+		// In Firefox, execCommand( 'insertText' ), needed to preserve the undo stack,
+		// always moves the cursor to the end of the content.
+		// A workaround involving manually setting the cursor position and inserting the editor content
+		// is needed to put the cursor back to the correct position.
+		textarea.value = newContent;
+		textarea.focus();
+		document.execCommand( 'insertText', false, content );
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		textarea.focus();
+		onInsert( newContent );
```
Comment (gwwar) [inline client/lib/html-toolbar/index.js]: Does anything bad happen if this is also the default action for Chrome?

File: client/lib/html-toolbar/index.js [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+	} else if ( !! Env.ie && Env.ie >= 11 ) {
+		// execCommand( 'insertText' ), needed to preserve the undo stack, does not exist in IE11.
+		// Using the previous version of replacing the entire content value instead.
+		textarea.value = newContent;
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		onInsert( newContent );
+	} else {
```
Comment (gwwar) [inline client/lib/html-toolbar/index.js]: Let's also try doing early returns over if/else block nesting.

File: client/my-sites/comments/comment/comment-html-editor.jsx [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+			this.setCommentContentValue
+		);
+		this.setState( ( { openTags } ) => ( {
+			openTags: isTagOpen
+				? openTags.concat( tag.name )
+				: openTags.filter( openTag => openTag !== tag.name ),
+		} ) );
+	};
+
+	onClickBold = () => this.insertHtmlTag( { name: 'strong' } );
```
Comment (gwwar) [inline client/my-sites/comments/comment/comment-html-editor.jsx]: Minor note, but naming wise, we prefer more descriptive names. Say `insertStrongTag` over `onClick` or `insertClosingTags` over `onClickCloseTags`

File: client/lib/html-toolbar/index.js (hunk 1 above)
Comment (Copons) [reply client/lib/html-toolbar/index.js]: Shoot!
I forgot we had split the comments bundle!
I'll see to sniff in another way. :)

File: client/lib/html-toolbar/index.js (hunk 2 above)
Comment (Copons) [reply client/lib/html-toolbar/index.js]: This is not "in sync" with the inline comment anymore.
For the post editor toolbar, I only needed to target IE11 to keep the undo history.
Here, though, there was another issue with the cursor positioning that doesn't work as expected on both FF and IE11+.
I believe it didn't happen on the post editor because, with TinyMCE around, there is plenty of stuff happening on content change, that also "accidentally" helped me with the toolbar.

So, in other words, we would not need to check for `insertText`, but for IE11 and Edge instead.

File: client/lib/html-toolbar/index.js (hunk 3 above)
Comment (Copons) [reply client/lib/html-toolbar/index.js]: Not bad, just unnecessary.
On Chrome `insertText` triggers the textarea's `onChange`, so there is no need to manually update the component state.

(Actually, but I haven't tested it, there might even be a duplication of the content insertion in the undo history 🤔)

File: client/my-sites/comments/comment/comment-html-editor.jsx (hunk 5 above)
Comment (Copons) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: I wanted to keep the naming as close to the post editor HTML toolbar (which, admittedly, are my bad as well 😅).
Though, as we move farther and farther from the post editor, I don't see any reason to keep this consistency in place. 🙂

File: client/lib/html-toolbar/index.js (hunk 1 above)
Comment (kwight) [reply client/lib/html-toolbar/index.js]: @gwwar oooooh, great reminder <3

File: client/lib/html-toolbar/index.js [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+		textarea.value = newContent;
+		textarea.focus();
+		document.execCommand( 'insertText', false, content );
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		textarea.focus();
+		onInsert( newContent );
+		return;
+	} else if ( /(?:edge|msie |trident.+?; rv:)/i.test( userAgent ) ) {
+		// execCommand( 'insertText' ), needed to preserve the undo stack, does not exist in IE11,
+		// and Edge has issues positioning the cursor after insertion.
```
Comment (gwwar) [inline client/lib/html-toolbar/index.js]: Could we update the comment here to explain why we're browser sniffing?

File: client/my-sites/comments/comment/comment-html-editor.jsx (hunk 5 above)
Comment (gwwar) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: Whatever feels best to you. As I mentioned before it's okay for us to diverge here, since I think we'll come up with different needs.

File: client/lib/html-toolbar/index.js (hunk 1 above)
Comment (gwwar) [reply client/lib/html-toolbar/index.js]: screen shot 2018-02-08 at 10 41 20 am

Hmm, still a 50% jump, but much better than +500%. Let's 👀 if this looks okay

File: client/my-sites/comments/comment/comment-html-editor.jsx [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { localize } from 'i18n-calypso';
+import { delay, map, reduce } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import AddImageDialog from 'post-editor/editor-html-toolbar/add-image-dialog';
```
Comment (gwwar) [inline client/my-sites/comments/comment/comment-html-editor.jsx]: Okay I think this is causing our bundle size to jump 50%. Can we avoid the dialogs for now?

Comment (gwwar) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: This toolbar component is probably a candidate for an async bundle

File: client/lib/html-toolbar/index.js [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertContent = ( textarea, before, content, after, onInsert = noop ) => {
+	const { selectionEnd, value } = textarea;
+	const newContent = before + content + after;
+	const userAgent = get( window, 'navigator.userAgent', '' );
+
+	if ( /(?:firefox|fxios)/i.test( userAgent ) ) {
+		// In Firefox, execCommand( 'insertText' ), needed to preserve the undo stack,
+		// always moves the cursor to the end of the content.
```
Comment (dmsnell) [inline client/lib/html-toolbar/index.js]: are we sure this is true when we have a selection already?

when I worked with it I though for sure that I could insert text without having the cursor jump to the very end. could it be only when no explicit selection is selected?

do we have a link to a bug report or a behavioral explanation we could add as a comment here?

File: client/my-sites/comments/comment/comment-html-editor.jsx (hunk 7 above)
Comment (gwwar) [reply client/my-sites/comments/comment/comment-html-editor.jsx]: As a quick test you can try commenting these out and see what the new bundle stats look like.

File: client/lib/html-toolbar/index.js [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+		onInsert( newContent );
+		return;
+	} else if ( /(?:edge|msie |trident.+?; rv:)/i.test( userAgent ) ) {
+		// execCommand( 'insertText' ), needed to preserve the undo stack, does not exist in IE11,
+		// and Edge has issues positioning the cursor after insertion.
+		textarea.value = newContent;
+		setCursorPosition( textarea, selectionEnd, newContent.length - value.length );
+		onInsert( newContent );
+		return;
+	}
```
Comment (dmsnell) [inline client/lib/html-toolbar/index.js]: keep in mind that we will never have a browser-change during execution of this code. if it's useful we can create separate browser-specific functions and once at boot-time detect the right ones to use then update some default behavior

```js
const insertTextVanilla = …
const insertTextForcePosition = …
const insertTextNoCommand = …

const insertText = (
	( isFirefox() && insertTextForcePosition ) ||
	( isIE() && insertTextNoCommand ) ||
	insertTextVanilla
);
```

Then we can skip all further calls and keep the implementations separated into their own functions.

File: client/lib/html-toolbar/index.js [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+ */
+export const attributesToString = ( attributes = {} ) =>
+	reduce(
+		attributes,
+		( attributesString, attributeValue, attributeName ) =>
+			attributeValue
+				? attributesString + ` ${ attributeName }="${ attributeValue }"`
+				: attributesString,
+		''
+	);
```
Comment (dmsnell) [inline client/lib/html-toolbar/index.js]: This algorithm suffers from the dangling-delimiter-dilemna - does the extra space go at the front or at the back?

```js
export const attributesToString = ( attributes = {} ) =>
	map( attributes, ( value, name ) => `${ name }="${ value }"` ).join( ' ' )
```

File: client/lib/html-toolbar/index.js [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+ * @param {Element} textarea The editor textarea.
+ * @param {Object} tag The tag to insert.
+ * @param {Function} onInsert A function to call after the insertion.
+ */
+export const insertHtmlTagOpenClose = ( textarea, tag, onInsert ) => {
+	const { before, inner, after } = splitSelectedContent( textarea );
+	insertContent(
+		textarea,
+		before,
+		openHtmlTag( tag ) + inner + closeHtmlTag( tag ),
```
Comment (dmsnell) [inline client/lib/html-toolbar/index.js]: with as sensitive as HTML can be and as easy as it is to break it we could also consider using the browser's built-in DOM methods here

```js
const l = m => console.log( m ) || m;
console.clear();

const makeElement = ( name, attributes, content ) => {
  const e = document.createElement( name )
  
  if ( attributes ) {
    Object
      .keys( attributes )
      .forEach( key => e.setAttribute( key, attributes[ key ] ) )
  }
  
  if ( content ) {
    e.innerHTML = content
  }
  
  return e
}

const makeHtml = ( ...args ) => makeElement( ...args ).outerHTML

// there's a better way…
const makeHtmlPair = ( name, attributes ) => {
  // the HTML comment _cannot_ exist inside a tag
  const e = makeElement( name, attributes, '<!---->' )
  
  return e.outerHTML.split( '<!---->' ) 
}
```

http://jsbin.com/getuhemule/1/edit?js,console

Comment (dmsnell) [reply client/lib/html-toolbar/index.js]: note for example that this takes HTML rules into play. I tried creating a pair for a `br` instead of a `p` and it only gave me one string back - because `br` _can't_ be split even though a simpler text-generation function can make it seem that way

**Comment by Copons** [comment]
Thanks @dmsnell for suggesting that approach.
I've moved this PR back to WIP to take a bit of time to work on all the comments.
Maybe trying to reuse the post editor HTML toolbar logic wasn't the best idea, and since we're building a new toolbar from scratch, we might as well build it better.

**Comment by Copons** [comment]
Replaced by https://github.com/Automattic/wp-calypso/pull/22331
